사장님 가게 수정

TIP-Santa opened the pull request:
#### 시나리오
---

1. 사장님 가게 수정
2. 사장님 가게 폐업

#### 테스트
---
- 성공 테스트
  - 수정 성공
  - 가게 폐업 성공
- 실패 테스트
  - 선택한 카테고리가 없는 경우
  - 동일한 가게이름이 존재하는 경우
  - 폐업인 가게가 영업으로 변경할 때 이미 영업중인 가게가 3개 이상인 경우

TIP-Santa committed:
#57 feat(StoreController) : patchStore 기능 추가

TIP-Santa committed:
#57 feat(PatchStoreRespDto) : responseDto 구성 추가

TIP-Santa committed:
#57 feat(PatchStoreReqDto) : requestDto 구성 추가

TIP-Santa committed:
#57 feat(StoreService) : 가게 정보 수정 기능 추가

TIP-Santa committed:
refactor(StoreCreateRespDto) : 에너테이션 추가

TIP-Santa committed:
refactor(StoreController) : Patch > Modify 이름 변경

TIP-Santa committed:
#57 refactor(StoreCategoryRepository) : 카테고리 수정을 위한 deleteByStoreId 추가

TIP-Santa committed:
chore(Store) : 필요없는 주석 제거

TIP-Santa committed:
#57 refactor
- ModifyStoreReqDto
- ModifyStoreRespDto
 : Patch > Modify 이름 변경

TIP-Santa committed:
#57 refactor(ModifyStoreReqDto) : isDeleted NotBlank > NotNull 수정

TIP-Santa committed:
#57 refactor(StoreService) : Builder 에 id 추가

TIP-Santa committed:
#57 refactor(StoreService) : 가게 정보 수정 기능 수정
- 가게 정보 수정 시 빈 데이터는 이전 데이터로
- 가게가 페업상태가 아닌 경우에만 카테고리 수정
- 가게 이름 중복 검사 시 name이 null이 아닌 경우에만 검사

TIP-Santa committed:
#57 test(StoreServiceTest) : 가게 수정 테스트
 - 성공 테스트
   - 정상 수정
   - 폐업
 - 실패 테스트
   - 선택한 카테고리가 존재하지 않는 경우
   - 동일한 가게 이름이 존재하는 경우
   - 이미 영업중인 가게가 3개이상 존재하는 경우

TIP-Santa committed:
dev pull > merge

TIP-Santa committed:
#57 test(StoreServiceTest) : 가게 수정 실패 테스트 - 최대 생성 수 제한 테스트의 전제 조건을 폐업중인 가게가 영업으로 변경시로 수정

JIMINJI1 reviewed (approved):
확인했습니다.

gbognon25 reviewed (commented):
확인했습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -55,6 +57,49 @@ public List<GetStoreRespDto> getAllStoresByOwnerId(Long ownerId) {
 
+    public ModifyStoreRespDto modifyStore(Long storeId, ModifyStoreReqDto modifyStoreReqDto, Long ownerId) {
+        // 가게 조회
+        Store store = storeRepository.findById(storeId).orElseThrow(() ->
+                new BusinessException(ErrorCode.STORE_NOT_FOUND)
+        );
```
조회 시 ownerId와 storeId를 같이 사용하는게 좋을 거 같습니다

TIP-Santa replied:
감사합니다.
ownerId를 확인하는 로직을 추가했습니다.
```
// 접속한 owner가 가게의 owner가 맞는지 확인
        if (!store.getOwner().getId().equals(ownerId)) {
            throw new BusinessException(ErrorCode.FORBIDDEN_ACCESS);
        }
```

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -55,6 +57,49 @@ public List<GetStoreRespDto> getAllStoresByOwnerId(Long ownerId) {
+        validateModifyStore(store, modifyStoreReqDto, ownerId);
+
+        // 가게 정보 수정: ModifyStoreReqDto의 값이 null이 아닌 경우만 업데이트
+        if (modifyStoreReqDto.getName() != null) {
+            store.setName(modifyStoreReqDto.getName());
+        }
+        if (modifyStoreReqDto.getDescription() != null) {
+            store.setDescription(modifyStoreReqDto.getDescription());
+        }
+        if (modifyStoreReqDto.getContact() != null) {
+            store.setContact(modifyStoreReqDto.getContact());
+        }
+        if (modifyStoreReqDto.getMinOrderAmount() != null) {
+            store.setMinOrderAmount(modifyStoreReqDto.getMinOrderAmount());
+        }
+        if (modifyStoreReqDto.getOpenTime() != null) {
+            store.setOpenTime(modifyStoreReqDto.getOpenTime());
+        }
+        if (modifyStoreReqDto.getEndTime() != null) {
+            store.setEndTime(modifyStoreReqDto.getEndTime());
+        }
+        if (modifyStoreReqDto.getIsDeleted() != null) {
+            store.setIsDeleted(modifyStoreReqDto.getIsDeleted());
+        }
```
dto의 모든 필드에 notblank, notnull이 지정되어 있어서 서비스단에서 null검증이 필요없습니다

현재는 수정 시 모든 필드 값을 보내게 되어있는데 구체적인 시나리오를 정하시면 좋을 거 같아요

TIP-Santa replied:
넵 dto를 수정하였습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/dto/req/ModifyStoreReqDto.java:
```diff
@@ -0,0 +1,42 @@
+@Setter
+@NoArgsConstructor
+public class ModifyStoreReqDto {
+    @NotBlank(message = "가게 이름를 입력해주세요")
+    @Size(min = 1, max = 20, message = "가게 이름은 최대 20자까지 입력할 수 있습니다.")
+    private String name;
+    @NotBlank(message = "가게 설명을 입력해주세요")
+    private String description;
+    @NotBlank(message = "가게의 전화번호를 입력해주세요")
+    @Size(min=1, max=15, message = "전화번호는 최대 15자까지 입력할 수 있습니다.")
+    private String contact;
+    @NotBlank(message = "최소 주문금액을 입력해주세요")
+    @Pattern(regexp = "^[0-9]+$", message = "최소 주문금액은 숫자만 입력할 수 있습니다.")
+    private Long minOrderAmount;
+    @NotBlank(message = "가게 오픈시간을 입력해주세요")
+    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "오픈시간은 시간으로 입력해주세요. ex) 11:00")
+    private LocalTime openTime;
+    @NotBlank(message = "가게 마감시간을 입력해주세요")
+    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "마감시간은 시간으로 입력해주세요. ex) 23:00")
+    private LocalTime endTime;
+    @NotBlank(message = "가게의 카테고리를 입력해주세요")
+    @Pattern(regexp = "^[0-9]+$", message = "카테고리에 맞는 숫자로 입력해주세요.")
+    private List<Long> categoryIds;
+    @NotNull(message = "가게의 영업상태를 설정해주세요")
+    private Boolean isDeleted;
```
모든 필드에 값을 받게 되어있네요

저희가 이전에 와이어프레임에서 이야기했던 대로라면 여기서의 notnull, notblank이 시나리오와 다릅니다ㅠ

현재는 프론트엔드에서 유저가 수정하지 않은 필드 값과 유저가 수정한 필드 값을 묶어서 수정 요청에 한꺼번에 보냈다는 시나리오를 가정하신 걸로 보이네요

이러한 시나리오를 의도하신 거라면 서비스 단에서의 null검증 없이 바로 도메인에 각 dto 필드 값을 넘겨서 수정하는 방식을 제안합니다 

Store 도메인에서 @Setter를 통한 방식이 아니라 modifyAll()과 같은 수정 메서드를 통해서요(@Setter는 제거해주세요!!)

TIP-Santa replied:
죄송합니다. 수정했습니다.
```
    @Size(min = 1, max = 20, message = "가게 이름은 최대 20자까지 입력할 수 있습니다.")
    private String name;
    private String description;
    @Size(min = 1, max = 15, message = "전화번호는 최대 15자까지 입력할 수 있습니다.")
    private String contact;
    @Pattern(regexp = "^[0-9]+$", message = "최소 주문금액은 숫자만 입력할 수 있습니다.")
    private Long minOrderAmount;
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "오픈시간은 시간으로 입력해주세요. ex) 11:00")
    private LocalTime openTime;
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "마감시간은 시간으로 입력해주세요. ex) 23:00")
    private LocalTime endTime;
    @Pattern(regexp = "^[0-9]+$", message = "카테고리에 맞는 숫자로 입력해주세요.")
    private List<Long> categoryIds;
    private Boolean isDeleted;
```

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -55,6 +57,49 @@ public List<GetStoreRespDto> getAllStoresByOwnerId(Long ownerId) {
                 .toList();
     }
 
+    public ModifyStoreRespDto modifyStore(Long storeId, ModifyStoreReqDto modifyStoreReqDto, Long ownerId) {
```
@Transactional 어노테이션이 추가되어야 더티 체킹이 수행됩니다

TIP-Santa replied:
앗 트랜젝션 자꾸 까먹네요 ㅠㅠ 추가했습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -55,6 +57,49 @@ public List<GetStoreRespDto> getAllStoresByOwnerId(Long ownerId) {
+        }
+
+        // 가게가 폐업 상태가 아닌 경우에만 카테고리 수정
+        if (modifyStoreReqDto.getIsDeleted() == null || !modifyStoreReqDto.getIsDeleted()) {
```
가게가 폐업 상태가 아닌 경우에 대해서만 수정을 진행하는 거라면 

위에서 storeId로 가게를 찾을 때 조건절에 isDeleted = false를 추가하시면 될 거 같습니다!

TIP-Santa replied:
영업중인 가게만 조회하게 되면 폐업 상태인 가게를 영업 상태로 수정할 수 없게 되어서 카테고리 수정 시에 가게의 페업 상태를 보고 카테고리를 수정할지 말지를 결정하는 게 더 좋을 것 같습니다.

yaewon0411 replied:
아 영업 상태도 같이 전환할 수 있네요 네네!!

yaewon0411 reviewed (commented):
테스트 통과 확인했습니다!


+)
storeServiceTest에 
    @DisplayName(("실패 테스트 - 존재하지 않는 카테고리 ID"))
    @Test
    public void testCreateStore_Fail_NotFoundCategoryId() 

이 테스트 스텁 설정이 하나 잘못되었던데 확인 바랍니다

TIP-Santa committed:
#57 refactor 리뷰 반영
- ModifyStoreReqDto : NotNull, NotBlank를 제거하여 수정이 필요하지 않는 경우 입력하지 않아도 되도록 변경하였습니다.
- ErrorCode : 수정된 내용이 존재하지 않을 경우 추가
- StoreService
   - @Transactional 어노테이션 추가
   - ownerId 확인 로직 추가
   - 수정된 데이터가 없을 경우 예외처리 추가

TIP-Santa committed:
#57 test 리뷰 반영
- StoreServiceTest
  - 필요업는 스텁 제거
  - 가게 수정 실패 테스트 추가
    : owner가 다른 경우
    : 수정된 내용이 없는 경우

TIP-Santa committed:
#57 chore(StoreService) : ErrorCode 이름 변경

TIP-Santa committed:
#57 chore(ErrorCode) : ErrorCode 이름 변경

yaewon0411 commented:
이 아래 예외는 어떤 이유에서 필요한가요?

```
        // 아무것도 수정되지 않은 경우 예외 처리
        if (!isModified) {
            throw new BusinessException(ErrorCode.NO_UPDATE_DATA);
        }
```

제가 봤을 땐 비즈니스 로직 상 실수로 변경 요청을 보냈을 때의 응답을 보내려고 하는 것 같은데 그렇다면 아래와 같은 방식을 제안합니다

```
@Getter
@Setter
@NoArgsConstructor
public class ModifyStoreReqDto {
    @Size(min = 1, max = 20, message = "가게 이름은 최대 20자까지 입력할 수 있습니다.")
    private String name;
    private String description;
    @Size(min = 1, max = 15, message = "전화번호는 최대 15자까지 입력할 수 있습니다.")
    private String contact;
    @Pattern(regexp = "^[0-9]+$", message = "최소 주문금액은 숫자만 입력할 수 있습니다.")
    private Long minOrderAmount;
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "오픈시간은 시간으로 입력해주세요. ex) 11:00")
    private LocalTime openTime;
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "마감시간은 시간으로 입력해주세요. ex) 23:00")
    private LocalTime endTime;
    @Pattern(regexp = "^[0-9]+$", message = "카테고리에 맞는 숫자로 입력해주세요.")
    private List<Long> categoryIds;
    private Boolean isDeleted;
    
    public boolean hasNoUpdateData() {
        return name == null && 
               description == null && 
               contact == null &&
               minOrderAmount == null && 
               openTime == null && 
               endTime == null && 
               isDeleted == null;
    }
}

```
```
@Transactional
    public ModifyStoreRespDto modifyStore(Long storeId, ModifyStoreReqDto dto, Long ownerId) {
       if (dto.hasNoUpdateData()) {
            throw new BusinessException("수정할 정보를 입력해주세요");
        }
        Store store = storeRepository.findById(storeId).orElseThrow(() ->
                new BusinessException(ErrorCode.STORE_NOT_FOUND)
        );
        store.updateAll(dto);  //서비스 단에 변경 로직이 너무 길어서 변경 기능은 해당 도메인에 위임하는 방식을 제안합니다
```

ex) 도메인 단에서 아래와 같이 null 체크 후 수정
(그냥 예시입니다  setter는 닫혀있는게 좋지만 일단 급하게 적느라 참고로만 봐주세요ㅠ)
```
public void updateAll(StoreUpdateDto dto) {
        Optional.ofNullable(dto.getName()).ifPresent(this::setName);
        Optional.ofNullable(dto.getDescription()).ifPresent(this::setDescription);
        Optional.ofNullable(dto.getContact()).ifPresent(this::setContact);
        Optional.ofNullable(dto.getMinOrderAmount()).ifPresent(this::setMinOrderAmount);
        Optional.ofNullable(dto.getOpenTime()).ifPresent(this::setOpenTime);
        Optional.ofNullable(dto.getEndTime()).ifPresent(this::setEndTime);
        Optional.ofNullable(dto.getIsDeleted()).ifPresent(this::setIsDeleted);
    }
```

```

TIP-Santa committed:
dev pull > merge

TIP-Santa committed:
#57 refactor 리뷰 반영
- StoreService
 : 가게 정보 수정 시 수정된 값이 없을 경우 처리 방법 변경
-  Store
 : 가게 정보 수정 관련 updateAll 추가
- ModifyStoreReqDto
 : 가게 정보 수정 관련 hasNoUpdateData 추가

TIP-Santa committed:
#57 test(StoreServiceTest) : 가게 수정 변경 관련 해당 테스트 코드 수정

TIP-Santa committed:
dev full > merge
